diff -Narup b/arch/x86/kernel/head64.c a/arch/x86/kernel/head64.c
--- b/arch/x86/kernel/head64.c	2024-04-03 09:32:51.000000000 -0400
+++ a/arch/x86/kernel/head64.c	2024-04-04 18:12:40.082440941 -0400
@@ -85,6 +85,9 @@ static struct desc_ptr startup_gdt_descr
 	.address = 0,
 };
 
+/* Required for read_cr3 when building as PIE */
+unsigned long __force_order;
+
 static void __head *fixup_pointer(void *ptr, unsigned long physaddr)
 {
 	return ptr - (void *)_text + (void *)physaddr;
diff -Narup b/arch/x86/kernel/Makefile fresh/arch/x86/kernel/Makefile
--- b/arch/x86/kernel/Makefile	2024-04-03 09:32:51.000000000 -0400
+++ a/arch/x86/kernel/Makefile	2024-04-04 18:12:04.822201150 -0400
@@ -21,6 +21,12 @@ CFLAGS_REMOVE_sev.o = -pg
 CFLAGS_REMOVE_rethook.o = -pg
 endif
 
+ifdef CONFIG_X86_PIE
+# Remove PIE and retpoline flags that are incompatible with mcmodel=large
+CFLAGS_REMOVE_head64.o += -fPIE -mindirect-branch=thunk-extern -mindirect-branch-register
+CFLAGS_head64.o = -mcmodel=large
+endif
+
 KASAN_SANITIZE_head$(BITS).o				:= n
 KASAN_SANITIZE_dumpstack.o				:= n
 KASAN_SANITIZE_dumpstack_$(BITS).o			:= n
