diff -ur a/arch/x86/boot/boot.h b/arch/x86/boot/boot.h
--- a/arch/x86/boot/boot.h	2024-04-03 21:28:19.383323217 -0400
+++ b/arch/x86/boot/boot.h	2024-04-03 09:32:51.000000000 -0400
@@ -168,7 +168,7 @@
 }
 
 /* Heap -- available for dynamic lists. */
-extern char _end[] __default_visibility;
+extern char _end[];
 extern char *HEAP;
 extern char *heap_end;
 #define RESET_HEAP() ((void *)( HEAP = _end ))
diff -ur a/arch/x86/include/asm/setup.h b/arch/x86/include/asm/setup.h
--- a/arch/x86/include/asm/setup.h	2024-04-03 21:29:24.975542348 -0400
+++ b/arch/x86/include/asm/setup.h	2024-04-03 09:32:51.000000000 -0400
@@ -73,7 +73,7 @@
  * This is set up by the setup-routine at boot-time
  */
 extern struct boot_params boot_params;
-extern char _text[] __default_visibility;
+extern char _text[];
 
 static inline bool kaslr_enabled(void)
 {
diff -ur a/include/asm-generic/sections.h b/include/asm-generic/sections.h
--- a/include/asm-generic/sections.h	2024-04-03 21:45:58.368384091 -0400
+++ b/include/asm-generic/sections.h	2024-04-03 09:32:51.000000000 -0400
@@ -32,10 +32,6 @@
  *	__softirqentry_text_start, __softirqentry_text_end
  *	__start_opd, __end_opd
  */
-#ifdef CONFIG_DEFAULT_HIDDEN_SYMS
-#pragma GCC visibility push(default)
-#endif
-
 extern char _text[], _stext[], _etext[];
 extern char _data[], _sdata[], _edata[];
 extern char __bss_start[], __bss_stop[];
@@ -53,9 +49,6 @@
 
 /* Start and end of .ctors section - used for constructor calls. */
 extern char __ctors_start[], __ctors_end[];
-#ifdef CONFIG_DEFAULT_HIDDEN_SYMS
-#pragma GCC visibility pop
-#endif
 
 /* Start and end of .opd section - used for function descriptors. */
 extern char __start_opd[], __end_opd[];
diff -ur a/include/linux/compiler.h b/include/linux/compiler.h
--- a/include/linux/compiler.h	2024-04-03 22:03:56.417500576 -0400
+++ b/include/linux/compiler.h	2024-04-03 09:32:51.000000000 -0400
@@ -199,13 +199,6 @@
 	__v;								\
 })
 
-#ifdef CONFIG_DEFAULT_HIDDEN_SYMS
-#pragma GCC visibility push(hidden)
-#define __default_visibility  __attribute__((visibility ("default")))
-#else
-#define __default_visibility
-#endif
-
 #endif /* __KERNEL__ */
 
 /*
diff -ur a/init/Kconfig b/init/Kconfig
--- a/init/Kconfig	2024-04-03 22:03:43.997399780 -0400
+++ b/init/Kconfig	2024-04-03 09:32:51.000000000 -0400
@@ -1941,13 +1941,6 @@
 
 source "kernel/Kconfig.kexec"
 
-#
-# Default to hidden visibility for all symbols.
-# Useful for Position Independent Code to reduce global references.
-#
-config DEFAULT_HIDDEN_SYMS
-	bool
-
 endmenu		# General setup
 
 source "arch/Kconfig"
diff -ur a/kernel/trace/trace.h b/kernel/trace/trace.h
--- a/kernel/trace/trace.h	2024-04-03 22:06:36.430770321 -0400
+++ b/kernel/trace/trace.h	2024-04-03 09:32:51.000000000 -0400
@@ -1886,8 +1886,8 @@
 extern int tracing_snapshot_cond_disable(struct trace_array *tr);
 extern void *tracing_cond_snapshot_data(struct trace_array *tr);
 
-extern const char *__start___trace_bprintk_fmt[] __default_visibility;
-extern const char *__stop___trace_bprintk_fmt[] __default_visibility;
+extern const char *__start___trace_bprintk_fmt[];
+extern const char *__stop___trace_bprintk_fmt[];
 
 extern const char *__start___tracepoint_str[];
 extern const char *__stop___tracepoint_str[];
diff -ur a/lib/dynamic_debug.c b/lib/dynamic_debug.c
--- a/lib/dynamic_debug.c	2024-04-03 22:09:30.196100319 -0400
+++ b/lib/dynamic_debug.c	2024-04-03 09:32:51.000000000 -0400
@@ -39,8 +39,8 @@
 
 #include <rdma/ib_verbs.h>
 
-extern struct _ddebug __start___dyndbg[] __default_visibility;
-extern struct _ddebug __stop___dyndbg[] __default_visibility;
+extern struct _ddebug __start___dyndbg[];
+extern struct _ddebug __stop___dyndbg[];
 extern struct ddebug_class_map __start___dyndbg_classes[];
 extern struct ddebug_class_map __stop___dyndbg_classes[];

