diff -Narup -W 500 a/arch/x86/kernel/kvm.c b/arch/x86/kernel/kvm.c
--- a/arch/x86/kernel/kvm.c	2024-04-25 18:04:31.290296531 -0400
+++ b/arch/x86/kernel/kvm.c	2024-04-25 22:49:15.300156971 -0400
@@ -800,8 +800,12 @@ extern bool __raw_callee_save___kvm_vcpu
  * restoring to/from the stack.
  */
 #define PV_VCPU_PREEMPTED_ASM						     \
- "movq   __per_cpu_offset(,%rdi,8), %rax\n\t"				     \
- "cmpb   $0, " __stringify(KVM_STEAL_TIME_preempted) "+steal_time(%rax)\n\t" \
+ "pushq	%rdi\n\t"																\
+ _ASM_LEA(__per_cpu_offset, %rip, %rax) "\n\t"									\
+ "movq	(%rax,%rdi,8), %rax\n\t"												\
+ "leaq	" __stringify(KVM_STEAL_TIME_preempted) "+steal_time(%rip), %rdi\n\t"	\
+ "cmpb	$0, (%rax,%rdi,1)\n\t"													\
+ "popq	%rdi\n\t"																\
  "setne  %al\n\t"
 
 DEFINE_ASM_FUNC(__raw_callee_save___kvm_vcpu_is_preempted,
